numba/array_analysis.py

```diff
@@ -1422,7 +1432,59 @@ def _analyze_op_call_numpy_reshape(self, scope, equiv_set, args, kws):
             typ = self.typemap[args[1].name]
             if isinstance(typ, types.BaseTuple):
                 return args[1], []
-        return tuple(args[1:]), []
+
+        # Reshape is allowed to take one argument that has the value -1.
+        # This means that the size of that dimension should be inferred from
+        # the size of the array being reshaped and the other dimensions
+        # specified.  Our general approach here is to see if the reshape
+        # has any -1 arguments.  If it has more than one then throw a
+        # ValueError.  If exactly one -1 argument is found, remember its
+        # argument index.
+        stmts = []
+        neg_one_index = -1
+        for arg_index in range(1, len(args)):
+            reshape_arg = args[arg_index]
+            reshape_arg_def = guard(get_definition, self.func_ir, reshape_arg)
+            if isinstance(reshape_arg_def, ir.Const):
+                if reshape_arg_def.value == -1:
+                    if neg_one_index == -1:
+                        neg_one_index = arg_index
+                    else:
+                        msg = ("The reshape API may only include one -1 argument.")
+                        raise errors.ApiError(msg, loc=reshape_arg.loc)
+
+        if neg_one_index >= 0:
+            # If exactly one -1 argument to reshape was found, then we are
+            # going to insert code to calculate the missing dimension and then
+            # replace the -1 with the calculated size.  We do this because we
+            # can't let array equivalence analysis think that some array has
+            # a -1 dimension size.
+            loc = args[0].loc
+            # Create a variable to hold the size of the array being reshaped.
+            calc_size_var = ir.Var(scope, mk_unique_var("calc_size_var"), loc)
+            self.typemap[calc_size_var.name] = types.intp
+            # Assign the size of the array calc_size_var.
+            init_calc_var = ir.Assign(ir.Expr.getattr(args[0], "size", loc), calc_size_var, loc)
+            stmts.append(init_calc_var)
+            # For each other dimension, divide the current size by the specified
+            # dimension size.  Once all such dimensions have been done then what is
+            # left is the size of the -1 dimension.
+            for arg_index in range(1, len(args)):
```

Comment:
can this be simplified by just doing the `unknown_dim = array.size // product(known_dims)` ?

Reply:
The known dims don't have to be contiguous so forming the product is just as much code as what I'm currently doing, right?

Reply:
Numba can't reshape a non-contiguous array?
```pycon
In [28]: from numba import njit                                                                                                 

In [29]: import numpy as np                                                                                                     

In [30]: a = np.zeros((6, 10))[::2]                                                                                             

In [31]: def foo(x): 
    ...:     return x.reshape(15, 2) 
    ...:                                                                                                                        

In [32]: foo(a)                                                                                                                 
Out[32]: 
array([[0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.]])

In [33]: njit(foo)(a) 
<traceback stuff>
numba.errors.InternalError: reshape() supports contiguous array only
[1] During: resolving callee type: BoundFunction(array.reshape for array(float64, 2d, A))
[2] During: typing of call at <ipython-input-31-f8ab8b7e296d> (2)

--%<----------------------------------------------------------------------------


File "<ipython-input-31-f8ab8b7e296d>", line 2:
def foo(x):
    return x.reshape(15, 2)
    ^
```
however, this is a good point and I guess makes it future proof so elect to leave it.